Terminal.Gui/View/ViewScrollBar.cs

```diff
@@ -0,0 +1,635 @@
+﻿namespace Terminal.Gui;
+
+/// <summary>
+///     The scroll bar types used by this <see cref="ScrollBarView"/>.
+/// </summary>
+public enum ScrollBarType
+{
+    /// <summary>
+    ///     None scroll bar will be used and to avoid throwing an exception never use it in a constructor.
+    /// </summary>
+    None,
+
+    /// <summary>
+    ///     Only the vertical scroll bar will be shown.
+    /// </summary>
+    Vertical,
+
+    /// <summary>
+    ///     Only the horizontal scroll bar will be shown.
+    /// </summary>
+    Horizontal,
+
+    /// <summary>
+    ///     Both vertical and horizontal scroll bars will be shown.
+    /// </summary>
+    Both
+}
+
+public partial class View
+{
+    private ScrollBarView _scrollBar;
+    private ScrollBarType _scrollBarType;
+    private int _scrollColsSize;
+    private int _scrollLeftOffset;
+    private int _scrollRowsSize;
+    private int _scrollTopOffset;
+
+    /// <summary>If true the vertical/horizontal scroll bars won't be showed if it's not needed.</summary>
+    public bool ScrollAutoHideScrollBars
```

Comment:
But I'm not exposing the `ScrollBar` in the `View` class. That's why the property exists in the View class.

Reply:
Right. I get it now.

I think the name of this property should be `AutoShowScrollBars` and it should be `false` by default.

When it gets set to `true` this logic should get executed (in `Padding`, IMO):

```cs
if (verticalShouldShow) {
  Parent.VerticalScrollBar.Visible = true;
}
if (horizShouldShow) {
  Parent.HoriztonalScrollBar.Visible = true;
}
```

When it gets set to `false` this logic should get executed (in `Padding`, IMO):

```cs
Parent.VerticalScrollBar.Visible = false;
Parent.HoriztonalScrollBar.Visible = false;
```

If either scrollbar has `Enabled == false` these calls will (and should) do nothing. Because, by definition of `Enabled`, the view effectively does not exist. This would be the case for a `View` where having scrollbars does not make sense (e.g. `Button`).

Reply:
But instead of `Enabled=false` why not `Visible=false`. Why you want to show a grayed `ScrollBar`?. This is really make to much confusion to me and I have others broken code to try to fix. If a `ScrollBar` doesn't make sense on a view I think is better force set on a override `Visible` to false.

Reply:
> But instead of `Enabled=false` why not `Visible=false`. Why you want to show a grayed `ScrollBar`?. This is really make to much confusion to me and I have others broken code to try to fix. If a `ScrollBar` doesn't make sense on a view I think is better force set on a override `Visible` to false.

I was not remembering what `Enabled` did correctly. My bad. See my comment above about needing two things:  A way for a View to say "I don't do scroll bars" and a way for scroll bars to be made visible/notvisible IF they're possible.

Reply:
Actually any view can have `ScrollBar`. Let the user choose what he want.